app/controllers/foreman_register/hosts_controller.rb

```diff
@@ -19,11 +19,20 @@ class HostsController < ::ForemanRegister::ApplicationController
     before_action :set_admin_user
     before_action :skip_secure_headers
     before_action :find_host
+    before_action :find_template
 
     def register
-      @host.setBuild
-      template = rendered_registration_template
-      render plain: template if template
+      if @template
+        @host.setBuild
```

Comment:
I am not comfortable with this to be honest. Is this really needed? Leos mentioned that you, @ares, wanted this.

This changes state of the host just by calling register endpoint which is supposed to render a template. This can either cause some orchestration issues (user-based hooks for example) or create an attack vector when an existing (managed) host could be put into build mode (possible data loss after reboot).

If we want the build mode, why don't we call set build explicitly? There is an API endpoint to do that.

Finally, there must was a reason why build validation for unmanaged hosts was added: https://projects.theforeman.org/issues/8587 specifically that the host is in incorrect state (PXE templates were not deployed). When a host is then managed, templates do not get rendered and it's still in incorrect mode.

Reply:
This PR basically allows unmanaged host to start using build mode signalling. Once the registration template is downloaded, it turns the host to "pending installation". After the registration is complete (can take minutes, as it can install puppet, insights etc), it turns it to "built". I think this is a great benefit comparing not knowing that some host is being registered. This tells you the host is not ready for using yet, even though it exists in the DB (we need to create it before we render the template for it)

> This changes state of the host just by calling register endpoint which is supposed to render a template. This can either cause some orchestration issues (user-based hooks for example)

I'm not concerned with that (in combination with the answer below). Let's assume we can only run the registration for unmanaged hosts. There's no orchestration for unmanaged hosts. I'm not sure about hooks plugin, but I don't think people would define any hooks for unmanaged hosts. They are typically not created from the Foreman, but rather by e.g. puppet.

> or create an attack vector when an existing (managed) host could be put into build mode (possible data loss after reboot).

You're right, we should make sure the `@host` is unmanaged, we shouldn't allow registering managed hosts.

> If we want the build mode, why don't we call set build explicitly? There is an API endpoint to do that.

we can't, the validation prevents unmanaged host to be set to be built, that's the change in this PR.

> Finally, there must was a reason why build validation for unmanaged hosts was added:

We should then perhaps fix this properly and reset the build mode when host is switched between managed - unmanaged. Or rerender the templates. The issue you linked was talking about possibility of setting the build mode for unmanaged host through API/UI. We allow this for unmanaged hosts only via registration process.

Reply:
Okay, just to confirm: the decision is to allow build mode **only for unmanaged hosts**

Reply:
I think the conclusion (I hope others agree with) is, keep removal of the validation so both managed and newly unmanaged host can be turned to build mode. But the `/register` endpoint should verify the specified host is actually unmanaged and fail if it is managed.

Reply:
As part of testing I figured the behavior described in here was the original intention and we should have this endpoint available for managed hosts too, calling register is similar to switching the host to build, it will clear facts and reports without a warning (through API), but that's by design. Only authorized users can do that.